fix(lsp): stop repeatedly resuming dead coroutine

ntdiary opened the pull request:
Problem:
Error extracting content-length causes all future coroutine resumes to
fail.

Solution:
Replace `coroutine.wrap` with `coroutine.create` in `create_read_loop`
so that we can check its status and catch any errors, allowing us to
stop the lsp client and avoid repeatedly resuming the dead coroutine.



close https://github.com/neovim/neovim/issues/33564

justinmk commented:
This looks rather specific (`get_content_length` is presumably not the only place where a BOM could cause problems in the LSP layer), but when problems like #33564 happen with unclear symptoms, we should at *minimum* detect and show useful errors. 

So even if there are objections to this PR as it currently is, at least detecting the BOM and showing a useful error is something we should do.

justinmk reviewed (approved):
Thanks for your analysis, can you include the key parts in your commit message? 

There is an existing test which uses `create_read_loop`, maybe it's not too hard to force a failure to exercise your changes in a test? https://github.com/neovim/neovim/blob/470585b5d063b374bc2efed14a7978b6eabf090c/test/functional/plugin/lsp_spec.lua#L5743

Ideally @mfussenegger can check this too. Will merge in a few days otherwise.

mfussenegger reviewed (approved):
I second https://github.com/neovim/neovim/pull/35743/files#r2462107498 and the request for a test case, otherwise looks okay to me.

I also wouldn't be opposed making the parsing a bit more lenient again and allow the first line to contain some garbage. E.g. one scenario I recently ran into where it also breaks is if you want to debug language servers written in Java with the jdwp agentlib attached. That will also cause an extra line to be printed, and there's no way to suppress that on the JDK/language server side.

ntdiary committed:
fix(lsp): stop repeatedly resuming dead coroutine

Problem:
Error extracting content-length causes all future coroutine resumes to
fail.

Solution:
Replace coroutine.wrap with coroutine.create in create_read_loop
so that we can check its status and catch any errors, allowing us to
stop the lsp client and avoid repeatedly resuming the dead coroutine.
